pkg/statistics/fmsketch.go

```diff
@@ -46,18 +46,29 @@ var fmSketchPool = sync.Pool{
 // TODO: add this attribute to PB and persist it instead of using a fixed number(executor.maxSketchSize)
 const MaxSketchSize = 10000
 
-// FMSketch (Flajolet–Martin Sketch) is a probabilistic data structure used for estimating the number of distinct elements in a stream.
-// It uses a hash function to map each element to a binary number and counts the number of trailing zeroes in each hashed value.
-// The maximum number of trailing zeroes observed gives an estimate of the logarithm of the number of distinct elements.
-// This approach allows the FM sketch to handle large streams of data in a memory-efficient way.
-//
-// See https://en.wikipedia.org/wiki/Flajolet%E2%80%93Martin_algorithm
+// FMSketch (Flajolet–Martin Sketch) is a probabilistic data structure that estimates the count of unique elements in a stream.
+// It employs a hash function to convert each element into a binary number and then counts the trailing zeroes in each hashed value.
+// **This variant of the FM sketch uses a set to store unique hashed values and a binary mask to track the maximum number of trailing zeroes.**
+// The estimated count of distinct values is calculated as 2^r * count, where 'r' is the maximum number of trailing zeroes observed and 'count' is the number of unique hashed values.
+// The underlying concept is that our hash function maps the input domain onto a logarithmic range. Each distinct value is mapped to 'i' with a probability of 2^-(i+1).
+// For example, a value is mapped to 0 with a probability of 1/2, to 1 with a probability of 1/4, to 2 with a probability of 1/8, and so on.
+// This is achieved by hashing the input value and counting the trailing zeroes in the hash value.
+// If we have a set of 'n' distinct values, the count of distinct values with 'r' trailing zeroes is n / 2^r.
+// Therefore, the estimated count of distinct values is 2^r * count = n.
+// The level-by-level approach increases the accuracy of the estimation by ensuring a minimum count of distinct values at each level.
+// This way, the final estimation is less likely to be skewed by outliers.
+// For more details, refer to the following papers:
+// 1. https://www.vldb.org/conf/2001/P541.pdf
```

Comment:
A numbered list marker is a decimal number of any length followed by a period or right parenthesis, then a space or tab, and then text. In a numbered list, each line beginning with a number list marker starts a new list item. Item numbers are left as is, never renumbered.


https://tip.golang.org/doc/comment

Reply:
Updated.